text/0000-dyno.md

```diff
@@ -0,0 +1,338 @@
+
+Beyond `Error`, one could imagine using the proposed API in any situation where we might add arbitrary data to a generic object. Another concrete example might be the `Context` object passed to `future::poll`.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In this section I'll describe how the `provide_any` module is used and defined. Most of the module is not intended to be used directly. Typically there is an intermediate library which has a trait which extends `Provider` (we will use the `Error` trait in our examples) and has API which is a facade over `provide_any`'s API. `provide_any` is relatively complex in both interface and implementation. However, most users will not even be aware of its existence. Using `provide_any` in an intermediate library has some complexity, but using the intermediate library is straightforward for end users.
+
+`provide_any` supports generic, type-driven access to data and a mechanism for intermediate implementers to provide such data. The key parts of `provide_any`'s interface are the `Provider` trait for objects which can provide data, and the `request_by_type_tag` function for requesting data from an object which implements `Provider`. Note that end users should not call `request_by_type_tag` directly, it is a helper function for intermediate implementers to use.
+
+For the rest of this section we will explain the earlier example, starting from data access and work our way down the implementation to `provide_any`. Note that the fact that `Error` is in the standard library is immaterial to its use of `provide_any`. `provide_any` is a public module and can be used by any crate. We use `Error` as an example of how `provide_any` is used, we are not proposing changes to `Error` in this RFC.
+
+A user of `Error` trait objects can call `get_context_ref` to access data by type which might be carried by an `Error` object. The function returns an `Option` and will return `None` if the requested type of data is not carried. For example, to access a backtrace (in a world where the `Error::backtrace` method has been removed): `e.get_context_ref::<Backtrace>()` (specifying the type using the turbofish syntax may not be necessary if the type can be inferred from the context).
+
+Lets examine the changes to `Error` required to make this work:
+
+```rust
+pub mod error {
+    pub trait Error: Debug + Provider {
+        ...
+        fn provide_context<'a>(&'a self, _req: Requisition<'a, '_>) {}
+    }
+
+    impl<T: Error> Provider for T {
+        fn provide<'a>(&'a self, req: Requisition<'a, '_>) {
```

Comment:
Is there a specific reason, why this is not `req: &mut Requisition`? Then it wouldn't be consumed and could be passed to different `provide` methods inside of one `Provider` (for example for composing different providers).

Reply:
That's a good question! It isn't because a `Requisition` is just a wrapper around an `&mut` pointer, so from the library's point of view it doesn't need to be. But you are right that the wrapper gives it move semantics rather than copying semantics and that might be inconvenient for the user here.

Reply:
Well, `&mut` doesn't have copy semantics either, what it have though is reborrows.